src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs

```diff
@@ -348,12 +401,143 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
+
+                        }
+                        ApplyResourceInNestedProperty(edmProperty, resource, resWrapper, readContext);
                     }
                 }
             }
         }
 
+
+        private ODataResourceSetWrapper CreateResourceSetWrapper(IEdmCollectionTypeReference edmPropertyType,
+    IList<ODataEntityReferenceLinkWrapper> refLinks, ODataDeserializerContext readContext)
+        {
+            ODataResourceSet resourceSet = new ODataResourceSet
+            {
+                TypeName = edmPropertyType.FullName(),
+            };
+
+            IEdmTypeReference elementType = edmPropertyType.ElementType();
+            ODataResourceSetWrapper resourceSetWrapper = new ODataResourceSetWrapper(resourceSet);
+            foreach (ODataEntityReferenceLinkWrapper refLinkWrapper in refLinks)
+            {
+                ODataResourceWrapper resourceWrapper = CreateResourceWrapper(elementType, refLinkWrapper, readContext);
+                resourceSetWrapper.Resources.Add(resourceWrapper);
+            }
+
+            return resourceSetWrapper;
+        }
+
+        private ODataResourceWrapper CreateResourceWrapper(IEdmTypeReference edmPropertyType, ODataEntityReferenceLinkWrapper refLink, ODataDeserializerContext readContext)
+        {
+            Contract.Assert(readContext != null);
+
+            ODataResource resource = new ODataResource
+            {
+                TypeName = edmPropertyType.FullName(),
+            };
+
+            Uri serviceRootUri = null;
+            string serviceRoot = "";
+            DefaultODataPathHandler pathHandler = new DefaultODataPathHandler();
+            if (!refLink.EntityReferenceLink.Url.IsAbsoluteUri)
+            {
+                IWebApiRequestMessage request1 = readContext.InternalRequest;
+                IWebApiUrlHelper urlHelper = readContext.InternalUrlHelper;            
+                serviceRoot = readContext.InternalUrlHelper.CreateODataLink(
+                    readContext.InternalRequest.Context.RouteName,
+                    readContext.InternalRequest.PathHandler,
+                    new List<ODataPathSegment>());
+
+                serviceRootUri = new Uri(serviceRoot, UriKind.Absolute);
+            }
+
+            var request = readContext.Request;
+            IEdmModel model = readContext.Model;
+
+            ODataPath odataPath = pathHandler.Parse(serviceRoot, refLink.EntityReferenceLink.Url.OriginalString,readContext.InternalRequest.RequestContainer);
+            KeySegment keySegment = odataPath.Segments.OfType<KeySegment>().LastOrDefault();
+            if (keySegment == null)
+            {
+                return null;
+            }
+
+            resource.Properties = keySegment.Keys.Select(k => new ODataProperty
+            {
+                Name = k.Key,
+                Value = k.Value
+            });
+
+            return new ODataResourceWrapper(resource);
```

Comment:
remember to copy the instance annotation from entity reference link to odata resource if any

Reply:
solved